File: src/handlers/callback.ts
/**
 * Callback query handler for Claude Telegram Bot.
 *
 * Handles inline keyboard button presses (ask_user MCP integration).
 */

import type { Context } from "grammy";
import { unlinkSync, readFileSync, existsSync } from "fs";
import { tmpdir } from "os";
import { resolve } from "path";
import { session } from "../session";
import { ALLOWED_USERS } from "../config";
import { isAuthorized } from "../security";
import { auditLog, sleep, startTypingIndicator } from "../utils";
import { StreamingState, createStatusCallback } from "./streaming";
import { parseRegistry } from "../registry";
import { GSD_OPERATIONS, parseRoadmap, handleGsd, handleProject, handleResume, handleRetry, sendGsdCommand } from "./commands";

/**
 * Handle callback queries from inline keyboards.
 */
export async function handleCallback(ctx: Context): Promise<void> {
  const userId = ctx.from?.id;
  const username = ctx.from?.username || "unknown";
  const chatId = ctx.chat?.id;
  const callbackData = ctx.callbackQuery?.data;

  if (!userId || !chatId || !callbackData) {
    await ctx.answerCallbackQuery();
    return;
  }

  // 1. Authorization check
  if (!isAuthorized(userId, ALLOWED_USERS)) {
    await ctx.answerCallbackQuery({ text: "Unauthorized" });
    return;
  }

  // 2. Handle resume callbacks: resume:{session_id}
  if (callbackData.startsWith("resume:")) {
    await handleResumeCallback(ctx, callbackData);
    return;
  }

  // 2b. Handle project callbacks: project:{index}
  if (callbackData.startsWith("project:")) {
    await handleProjectCallback(ctx, callbackData);
    return;
  }

  // 2c. Handle quick action callbacks: action:{name}
  if (callbackData.startsWith("action:")) {
    await handleActionCallback(ctx, callbackData, chatId);
    return;
  }

  // 2d. Handle gsd-run: — Run a GSD command in the current session
  if (callbackData.startsWith("gsd-run:")) {
    const command = callbackData.replace("gsd-run:", "");
    try { await ctx.deleteMessage(); } catch {}
    await ctx.answerCallbackQuery();
    await sendGsdCommand(ctx, command, command, username, userId, chatId);
    return;
  }

  // 2e. Handle gsd-fresh: — Clear session, then run GSD command
  if (callbackData.startsWith("gsd-fresh:")) {
    const command = callbackData.replace("gsd-fresh:", "");
    try { await ctx.deleteMessage(); } catch {}
    await ctx.answerCallbackQuery();
    // Kill session first
    if (session.isRunning) {
      await session.stop();
      await sleep(100);
      session.clearStopRequested();
    }
    await session.kill();
    await sendGsdCommand(ctx, command, command, username, userId, chatId);
    return;
  }

  // 2f. Handle numbered option callbacks: option:{number}
  if (callbackData.startsWith("option:")) {
    await handleOptionCallback(ctx, callbackData, chatId);
    return;
  }

  // 2g. Handle GSD callbacks: gsd:{operation}
  if (callbackData.startsWith("gsd:")) {
    await handleGsdCallback(ctx, callbackData, chatId);
    return;
  }

  // 2h. Handle GSD phase picker: gsd-{op}:{phase}
  const gsdPhasePrefixes = ["gsd-exec:", "gsd-plan:", "gsd-discuss:", "gsd-research:", "gsd-verify:", "gsd-remove:"];
  if (gsdPhasePrefixes.some((p) => callbackData.startsWith(p))) {
    await handleGsdPhaseCallback(ctx, callbackData, chatId);
    return;
  }

  // 3. Parse callback data: askuser:{request_id}:{option_index}
  if (!callbackData.startsWith("askuser:")) {
    await ctx.answerCallbackQuery();
    return;
  }

  const parts = callbackData.split(":");
  if (parts.length !== 3) {
    await ctx.answerCallbackQuery({ text: "Invalid callback data" });
    return;
  }

  const requestId = parts[1]!;
  const optionIndex = parseInt(parts[2]!, 10);

  // 3. Load request file
  const requestFile = resolve(tmpdir(), `ask-user-${requestId}.json`);
  let requestData: {
    question: string;
    options: string[];
    status: string;
  };

  try {
    const text = readFileSync(requestFile, "utf-8");
    requestData = JSON.parse(text);
  } catch (error) {
    console.error(`Failed to load ask-user request ${requestId}:`, error);
    await ctx.answerCallbackQuery({ text: "Request expired or invalid" });
    return;
  }

  // 4. Get selected option
  if (optionIndex < 0 || optionIndex >= requestData.options.length) {
    await ctx.answerCallbackQuery({ text: "Invalid option" });
    return;
  }

  const selectedOption = requestData.options[optionIndex]!;

  // 5. Update the message to show selection
  try {
    await ctx.editMessageText(`✓ ${selectedOption}`);
  } catch (error) {
    console.debug("Failed to edit callback message:", error);
  }

  // 6. Answer the callback
  await ctx.answerCallbackQuery({
    text: `Selected: ${selectedOption.slice(0, 50)}`,
  });

  // 7. Delete request file
  try {
    unlinkSync(requestFile);
  } catch (error) {
    console.debug("Failed to delete request file:", error);
  }

  // 8. Send the choice to Claude as a message
  const message = selectedOption;

  // Interrupt any running query - button responses are always immediate
  if (session.isRunning) {
    console.log("Interrupting current query for button response");
    await session.stop();
    // Small delay to ensure clean interruption
    await new Promise((resolve) => setTimeout(resolve, 100));
  }

  // Start typing
  const typing = startTypingIndicator(ctx);

  // Create streaming state
  const state = new StreamingState();
  const statusCallback = createStatusCallback(ctx, state);

  try {
    const response = await session.sendMessageStreaming(
      message,
      username,
      userId,
      statusCallback,
      chatId,
      ctx
    );

    await auditLog(userId, username, "CALLBACK", message, response);
  } catch (error) {
    console.error("Error processing callback:", error);

    for (const toolMsg of state.toolMessages) {
      try {
        await ctx.api.deleteMessage(toolMsg.chat.id, toolMsg.message_id);
      } catch (error) {
        console.debug("Failed to delete tool message:", error);
      }
    }

    if (String(error).includes("abort") || String(error).includes("cancel")) {
      // Only show "Query stopped" if it was an explicit stop, not an interrupt from a new message
      const wasInterrupt = session.consumeInterruptFlag();
      if (!wasInterrupt) {
        await ctx.reply("🛑 Query stopped.");
      }
    } else {
      await ctx.reply(`❌ Error: ${String(error).slice(0, 200)}`);
    }
  } finally {
    typing.stop();
  }
}

/**
 * Handle quick action button callbacks (action:{name}).
 * Dismisses the button row and delegates to the corresponding command handler.
 */
async function handleActionCallback(
  ctx: Context,
  callbackData: string,
  chatId: number
): Promise<void> {
  const action = callbackData.replace("action:", "");

  // Remove the button row message
  try {
    await ctx.deleteMessage();
  } catch {
    // May already be deleted
  }
  await ctx.answerCallbackQuery();

  switch (action) {
    case "stop":
      if (session.isRunning) {
        const result = await session.stop();
        if (result) {
          await sleep(100);
          session.clearStopRequested();
        }
        await ctx.reply("🛑 Stopped.");
      } else {
        await ctx.reply("Nothing running.");
      }
      break;

    case "retry":
      await handleRetry(ctx);
      break;

    case "new":
      if (session.isRunning) {
        await session.stop();
        await sleep(100);
        session.clearStopRequested();
      }
      await session.kill();
      await ctx.reply("🆕 Session cleared. Next message starts fresh.");
      break;

    case "gsd":
      await handleGsd(ctx);
      break;

    case "project":
      await handleProject(ctx);
      break;

    case "resume":
      await handleResume(ctx);
      break;

    default:
      await ctx.reply(`Unknown action: ${action}`);
  }
}

/**
 * Handle resume session callback (resume:{session_id}).
 */
async function handleResumeCallback(
  ctx: Context,
  callbackData: string
): Promise<void> {
  const userId = ctx.from?.id;
  const username = ctx.from?.username || "unknown";
  const chatId = ctx.chat?.id;
  const sessionId = callbackData.replace("resume:", "");

  if (!sessionId || !userId || !chatId) {
    await ctx.answerCallbackQuery({ text: "Invalid session ID" });
    return;
  }

  // Check if session is already active
  if (session.isActive) {
    await ctx.answerCallbackQuery({ text: "Session already active" });
    return;
  }

  // Resume the selected session
  const [success, message] = session.resumeSession(sessionId);

  if (!success) {
    await ctx.answerCallbackQuery({ text: message, show_alert: true });
    return;
  }

  // Update the original message to show selection
  try {
    await ctx.editMessageText(`✅ ${message}`);
  } catch (error) {
    console.debug("Failed to edit resume message:", error);
  }
  await ctx.answerCallbackQuery({ text: "Session resumed!" });

  // Send a hidden recap prompt to Claude
  const recapPrompt =
    "Please write a very concise recap of where we are in this conversation, to refresh my memory. Max 2-3 sentences.";

  const typing = startTypingIndicator(ctx);
  const state = new StreamingState();
  const statusCallback = createStatusCallback(ctx, state);

  try {
    await session.sendMessageStreaming(
      recapPrompt,
      username,
      userId,
      statusCallback,
      chatId,
      ctx
    );
  } catch (error) {
    console.error("Error getting recap:", error);
    // Don't show error to user - session is still resumed, recap just failed
  } finally {
    typing.stop();
  }
}

/**
 * Handle project switch callback (project:{index}).
 */
async function handleProjectCallback(
  ctx: Context,
  callbackData: string
): Promise<void> {
  const index = parseInt(callbackData.replace("project:", ""), 10);

  // Re-parse registry with same sort order
  const projects = parseRegistry();

  if (isNaN(index) || index < 0 || index >= projects.length) {
    await ctx.answerCallbackQuery({ text: "Invalid project selection" });
    return;
  }

  const project = projects[index]!;

  // Normalize path for Windows (registry uses backslashes)
  const projectPath = project.location.replace(/\//g, "\\");

  // Validate directory exists
  if (!existsSync(projectPath)) {
    await ctx.answerCallbackQuery({
      text: `Directory not found: ${projectPath}`,
      show_alert: true,
    });
    return;
  }

  // Kill active session if any
  if (session.isRunning) {
    await session.stop();
    await new Promise((r) => setTimeout(r, 100));
    session.clearStopRequested();
  }
  if (session.isActive) {
    await session.kill();
  }

  // Switch working directory
  try {
    session.setWorkingDir(projectPath);
  } catch (error) {
    await ctx.answerCallbackQuery({
      text: `Failed: ${String(error).slice(0, 100)}`,
      show_alert: true,
    });
    return;
  }

  // Update message to show confirmation
  try {
    await ctx.editMessageText(
      `📂 Switched to <b>${project.name}</b>\n<code>${projectPath}</code>`,
      { parse_mode: "HTML" }
    );
  } catch (error) {
    console.debug("Failed to edit project message:", error);
  }

  await ctx.answerCallbackQuery({ text: `Switched to ${project.name}` });
}

/**
 * Handle numbered option callback (option:{number}).
 * Sends the option number to Claude as text input.
 */
async function handleOptionCallback(
  ctx: Context,
  callbackData: string,
  chatId: number
): Promise<void> {
  const username = ctx.from?.username || "unknown";
  const userId = ctx.from?.id!;
  const optNum = callbackData.replace("option:", "");

  // Delete the action bar (buttons become stale after selection)
  try {
    await ctx.deleteMessage();
  } catch {}
  await ctx.answerCallbackQuery({ text: `Selected option ${optNum}` });

  // Send the number to Claude as text (same as user typing "1")
  await sendGsdCommand(ctx, optNum, `Option ${optNum}`, username, userId, chatId);
}

/**
 * Operations that show a phase picker instead of sending immediately.
 */
const PHASE_PICKER_OPS: Record<string, string> = {
  execute: "gsd-exec",
  plan: "gsd-plan",
  discuss: "gsd-discuss",
  research: "gsd-research",
  verify: "gsd-verify",
  "remove-phase": "gsd-remove",
};

/**
 * Handle GSD operation callback (gsd:{operation}).
 * Some operations show a phase picker, others send immediately.
 */
async function handleGsdCallback(
  ctx: Context,
  callbackData: string,
  chatId: number
): Promise<void> {
  const username = ctx.from?.username || "unknown";
  const userId = ctx.from?.id!;
  const operation = callbackData.replace("gsd:", "");

  // Find the matching operation
  const entry = GSD_OPERATIONS.find((op) => op[0] === operation);
  if (!entry) {
    await ctx.answerCallbackQuery({ text: "Unknown GSD operation" });
    return;
  }

  const [, label, command] = entry;

  // Check if this operation needs a phase picker
  const pickerPrefix = PHASE_PICKER_OPS[operation];
  if (pickerPrefix) {
    const phases = parseRoadmap(session.currentWorkingDir);
    // Verify shows all phases (done + pending), others show only pending
    const showDone = operation === "verify";
    const filteredPhases = showDone
      ? phases.filter((p) => p.status !== "skipped")
      : phases.filter((p) => p.status === "pending");

    if (filteredPhases.length === 0) {
      await ctx.answerCallbackQuery({
        text: showDone ? "No phases found" : "No pending phases found",
        show_alert: true,
      });
      return;
    }

    // Build phase picker keyboard: one button per row
    const buttons = filteredPhases.map((p) => [
      {
        text: `Phase ${p.number}: ${p.name}`,
        callback_data: `${pickerPrefix}:${p.number}`,
      },
    ]);

    try {
      await ctx.editMessageText(
        `<b>GSD</b> → ${label}\n\nSelect a phase:`,
        {
          parse_mode: "HTML",
          reply_markup: { inline_keyboard: buttons },
        }
      );
    } catch (error) {
      console.debug("Failed to edit GSD message:", error);
    }

    await ctx.answerCallbackQuery();
    return;
  }

  // Direct operations: send command immediately
  try {
    await ctx.editMessageText(`<b>GSD</b> → ${label}`, {
      parse_mode: "HTML",
    });
  } catch (error) {
    console.debug("Failed to edit GSD message:", error);
  }

  await ctx.answerCallbackQuery({ text: label });

  await sendGsdCommand(ctx, command, label, username, userId, chatId);
}

/**
 * Map gsd phase callback prefix to slash command and label.
 */
const PHASE_CALLBACK_MAP: Record<string, [string, string]> = {
  "gsd-exec": ["/gsd:execute-phase", "Execute Phase"],
  "gsd-plan": ["/gsd:plan-phase", "Plan Phase"],
  "gsd-discuss": ["/gsd:discuss-phase", "Discuss Phase"],
  "gsd-research": ["/gsd:research-phase", "Research Phase"],
  "gsd-verify": ["/gsd:verify-work", "Verify Phase"],
  "gsd-remove": ["/gsd:remove-phase", "Remove Phase"],
};

/**
 * Handle GSD phase picker callback (gsd-{op}:{phase}).
 */
async function handleGsdPhaseCallback(
  ctx: Context,
  callbackData: string,
  chatId: number
): Promise<void> {
  const username = ctx.from?.username || "unknown";
  const userId = ctx.from?.id!;

  const [prefix, phaseNum] = callbackData.split(":");
  const mapping = PHASE_CALLBACK_MAP[prefix!];
  if (!mapping || !phaseNum) {
    await ctx.answerCallbackQuery({ text: "Unknown phase operation" });
    return;
  }
  const [slashCmd, labelPrefix] = mapping;
  const command = `${slashCmd} ${phaseNum}`;
  const label = `${labelPrefix} ${phaseNum}`;

  // Update message to show selection
  try {
    await ctx.editMessageText(`<b>GSD</b> → ${label}`, {
      parse_mode: "HTML",
    });
  } catch (error) {
    console.debug("Failed to edit GSD phase message:", error);
  }

  await ctx.answerCallbackQuery({ text: label });

  await sendGsdCommand(ctx, command, label, username, userId, chatId);
}

File: internal/handlers/callback.go
package handlers

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"

	"github.com/PaulSonOfLars/gotgbot/v2"
	"github.com/PaulSonOfLars/gotgbot/v2/ext"

	"github.com/user/gsd-tele-go/internal/claude"
	"github.com/user/gsd-tele-go/internal/config"
	"github.com/user/gsd-tele-go/internal/project"
	"github.com/user/gsd-tele-go/internal/session"
)

// callbackAction is an enum for the action to take after parsing callback data.
// Using a pure enum lets us test routing logic without gotgbot types.
type callbackAction int

const (
	callbackActionUnknown       callbackAction = iota
	callbackActionResume                       // data = "resume:<session_id>"
	callbackActionStop                         // data = "action:stop"
	callbackActionNew                          // data = "action:new"
	callbackActionRetry                        // data = "action:retry"
	callbackActionGsd                          // data = "gsd:{operation_key}"
	callbackActionGsdRun                       // data = "gsd-run:{command}"
	callbackActionGsdFresh                     // data = "gsd-fresh:{command}"
	callbackActionGsdPhase                     // data = "gsd-exec:{N}", "gsd-plan:{N}", etc.
	callbackActionOption                       // data = "option:{key}"
	callbackActionAskUser                      // data = "askuser:{request_id}:{option_index}"
	callbackActionProjectChange                // data = "project:change"
	callbackActionProjectUnlink                // data = "project:unlink"
)

// parseCallbackData parses the callback data string and returns the action and
// optional payload (e.g. session ID for resume).
// Pure function — no Telegram types — fully testable.
//
// IMPORTANT: gsd-run:, gsd-fresh:, gsd-exec:, gsd-plan:, gsd-discuss:, gsd-research:,
// gsd-verify:, gsd-remove: are all checked BEFORE "gsd:" to prevent the shorter prefix
// from matching first.
func parseCallbackData(data string) (callbackAction, string) {
	switch {
	case strings.HasPrefix(data, "resume:"):
		return callbackActionResume, strings.TrimPrefix(data, "resume:")
	case data == "action:stop":
		return callbackActionStop, ""
	case data == "action:new":
		return callbackActionNew, ""
	case data == "action:retry":
		return callbackActionRetry, ""

	// GSD sub-prefixes MUST come before "gsd:" to avoid premature match.
	case strings.HasPrefix(data, "gsd-run:"):
		return callbackActionGsdRun, strings.TrimPrefix(data, "gsd-run:")
	case strings.HasPrefix(data, "gsd-fresh:"):
		return callbackActionGsdFresh, strings.TrimPrefix(data, "gsd-fresh:")
	case strings.HasPrefix(data, "gsd-exec:"),
		strings.HasPrefix(data, "gsd-plan:"),
		strings.HasPrefix(data, "gsd-discuss:"),
		strings.HasPrefix(data, "gsd-research:"),
		strings.HasPrefix(data, "gsd-verify:"),
		strings.HasPrefix(data, "gsd-remove:"):
		return callbackActionGsdPhase, data // keep full data for prefix+number parsing

	case strings.HasPrefix(data, "gsd:"):
		return callbackActionGsd, strings.TrimPrefix(data, "gsd:")

	case strings.HasPrefix(data, "option:"):
		return callbackActionOption, strings.TrimPrefix(data, "option:")
	case strings.HasPrefix(data, "askuser:"):
		return callbackActionAskUser, strings.TrimPrefix(data, "askuser:")
	case data == "project:change":
		return callbackActionProjectChange, ""
	case data == "project:unlink":
		return callbackActionProjectUnlink, ""

	default:
		return callbackActionUnknown, ""
	}
}

// callbackWg is a no-op WaitGroup used by callback handlers.
// Callbacks only enqueue to already-running workers; they do not start new ones.
// The bot-level WaitGroup tracks workers started from HandleText/restoreSessions.
var callbackWg sync.WaitGroup

// HandleCallback handles all inline keyboard callback queries.
//
// Routes by callback data prefix:
//   - "resume:<session_id>"     — restore a persisted Claude session
//   - "action:stop"             — stop the current running query
//   - "action:new"              — start a fresh session
//   - "action:retry"            — placeholder (deferred to v2)
//   - "gsd:{key}"               — run or pick a GSD operation
//   - "gsd-run:{cmd}"           — run a GSD command in current session
//   - "gsd-fresh:{cmd}"         — run a GSD command in a new session
//   - "gsd-exec:{N}" etc.       — run a phase-specific GSD command
//   - "option:{key}"            — send a numbered/lettered option to Claude
//   - "askuser:{id}:{idx}"      — respond to an ask_user MCP request
//   - "project:change"          — prompt user for new project path
//   - "project:unlink"          — unlink current project
func HandleCallback(b *gotgbot.Bot, ctx *ext.Context, store *session.SessionStore,
	persist *session.PersistenceManager, cfg *config.Config,
	mappings *project.MappingStore, awaitingPath *AwaitingPathState) error {

	if ctx.CallbackQuery == nil {
		return nil
	}

	cq := ctx.CallbackQuery

	// Answer the callback query immediately to remove the loading spinner.
	if _, err := cq.Answer(b, nil); err != nil {
		// Non-fatal: the user experience degrades slightly but we continue.
		_ = err
	}

	// Determine the chat ID from the message attached to the callback query.
	var chatID int64
	var msgID int64
	if cq.Message != nil {
		chat := cq.Message.GetChat()
		chatID = chat.Id
		msgID = cq.Message.GetMessageId()
	}

	if chatID == 0 {
		return nil
	}

	action, payload := parseCallbackData(cq.Data)

	switch action {
	case callbackActionResume:
		return handleCallbackResume(b, ctx, store, cfg, chatID, msgID, payload)

	case callbackActionStop:
		return handleCallbackStop(b, store, chatID, msgID)

	case callbackActionNew:
		return handleCallbackNew(b, store, cfg, chatID, msgID)

	case callbackActionRetry:
		// Deferred to v2.
		_, err := b.SendMessage(chatID, "Retry not available yet.", nil)
		return err

	case callbackActionGsd:
		return handleCallbackGsd(b, chatID, msgID, payload, store, mappings, cfg)

	case callbackActionGsdRun:
		if msgID != 0 {
			_, _, _ = b.EditMessageText("Running: "+payload+"...", &gotgbot.EditMessageTextOpts{
				ChatId:    chatID,
				MessageId: msgID,
			})
		}
		return enqueueGsdCommand(b, chatID, payload, store, mappings, cfg)

	case callbackActionGsdFresh:
		// Fresh session: clear session ID before enqueueing.
		if sess, ok := store.Get(chatID); ok {
			sess.SetSessionID("")
		}
		if msgID != 0 {
			_, _, _ = b.EditMessageText("Fresh session: "+payload+"...", &gotgbot.EditMessageTextOpts{
				ChatId:    chatID,
				MessageId: msgID,
			})
		}
		return enqueueGsdCommand(b, chatID, payload, store, mappings, cfg)

	case callbackActionGsdPhase:
		return handleCallbackGsdPhase(b, chatID, msgID, payload, store, mappings, cfg)

	case callbackActionOption:
		// Send the option key (e.g. "1", "A") directly to Claude.
		if msgID != 0 {
			_, _, _ = b.EditMessageText("Selected: "+payload, &gotgbot.EditMessageTextOpts{
				ChatId:    chatID,
				MessageId: msgID,
			})
		}
		return enqueueGsdCommand(b, chatID, payload, store, mappings, cfg)

	case callbackActionAskUser:
		return handleCallbackAskUser(b, chatID, msgID, payload, store, mappings, cfg)

	case callbackActionProjectChange:
		awaitingPath.Set(chatID)
		_, err := b.SendMessage(chatID, "Send the new project path:", nil)
		return err

	case callbackActionProjectUnlink:
		_ = mappings.Remove(chatID)
		// Stop the session if running.
		if sess, ok := store.Get(chatID); ok && sess.IsRunning() {
			sess.Stop()
		}
		if msgID != 0 {
			_, _, _ = b.EditMessageText("Project unlinked.", &gotgbot.EditMessageTextOpts{
				ChatId:    chatID,
				MessageId: msgID,
			})
		} else {
			_, _ = b.SendMessage(chatID, "Project unlinked.", nil)
		}
		return nil

	default:
		// Unknown callback data — log and ignore gracefully.
		return nil
	}
}

// handleCallbackGsd routes a "gsd:{key}" callback.
// If the key is in PhasePickerOps, show the phase picker keyboard.
// Otherwise, send the GSD command to Claude.
func handleCallbackGsd(b *gotgbot.Bot, chatID, msgID int64, key string,
	store *session.SessionStore, mappings *project.MappingStore,
	cfg *config.Config) error {

	// Check if this operation needs a phase picker first.
	if prefix, ok := PhasePickerOps[key]; ok {
		mapping, hasMapped := mappings.Get(chatID)
		if !hasMapped {
			_, err := b.SendMessage(chatID, "No project linked. Use /project to link one first.", nil)
			return err
		}
		phases := ParseRoadmap(mapping.Path)
		keyboard := BuildPhasePickerKeyboard(phases, prefix)
		_, err := b.SendMessage(chatID, "Select a phase:", &gotgbot.SendMessageOpts{
			ReplyMarkup: keyboard,
		})
		return err
	}

	// Look up the operation command.
	op, ok := gsdOpIndex[key]
	if !ok {
		_, err := b.SendMessage(chatID, "Unknown GSD operation: "+key, nil)
		return err
	}

	if msgID != 0 {
		_, _, _ = b.EditMessageText("Running: "+op.Command+"...", &gotgbot.EditMessageTextOpts{
			ChatId:    chatID,
			MessageId: msgID,
		})
	}
	return enqueueGsdCommand(b, chatID, op.Command, store, mappings, cfg)
}

// handleCallbackGsdPhase parses a phase-specific callback like "gsd-exec:2"
// and maps it to the full command "/gsd:execute-phase 2".
func handleCallbackGsdPhase(b *gotgbot.Bot, chatID, msgID int64, data string,
	store *session.SessionStore, mappings *project.MappingStore,
	cfg *config.Config) error {

	// Map callback prefix to GSD command.
	prefixToCmd := map[string]string{
		"gsd-exec":     "/gsd:execute-phase",
		"gsd-plan":     "/gsd:plan-phase",
		"gsd-discuss":  "/gsd:discuss-phase",
		"gsd-research": "/gsd:research-phase",
		"gsd-verify":   "/gsd:verify-work",
		"gsd-remove":   "/gsd:remove-phase",
	}

	// Split "gsd-exec:2" into prefix="gsd-exec" and number="2".
	colonIdx := strings.Index(data, ":")
	if colonIdx < 0 {
		_, err := b.SendMessage(chatID, "Invalid phase callback: "+data, nil)
		return err
	}
	prefix := data[:colonIdx]
	phaseNum := data[colonIdx+1:]

	cmd, ok := prefixToCmd[prefix]
	if !ok {
		_, err := b.SendMessage(chatID, "Unknown phase callback prefix: "+prefix, nil)
		return err
	}

	fullCmd := cmd + " " + phaseNum
	if msgID != 0 {
		_, _, _ = b.EditMessageText("Running: "+fullCmd+"...", &gotgbot.EditMessageTextOpts{
			ChatId:    chatID,
			MessageId: msgID,
		})
	}
	return enqueueGsdCommand(b, chatID, fullCmd, store, mappings, cfg)
}

// askUserRequest is the JSON structure written by the ask_user MCP tool.
type askUserRequest struct {
	Question string   `json:"question"`
	Options  []string `json:"options"`
	Status   string   `json:"status"`
}

// handleCallbackAskUser handles an "askuser:{request_id}:{option_index}" callback.
// It reads the temp file, validates the selection, edits the message, deletes the file,
// and sends the selected option text to Claude.
func handleCallbackAskUser(b *gotgbot.Bot, chatID, msgID int64, payload string,
	store *session.SessionStore, mappings *project.MappingStore,
	cfg *config.Config) error {

	// Payload format: "{request_id}:{option_index}"
	colonIdx := strings.Index(payload, ":")
	if colonIdx < 0 {
		_, err := b.SendMessage(chatID, "Invalid askuser payload.", nil)
		return err
	}
	requestID := payload[:colonIdx]
	optionIndexStr := payload[colonIdx+1:]

	// Parse option index.
	optionIndex := 0
	for _, ch := range optionIndexStr {
		if ch < '0' || ch > '9' {
			_, err := b.SendMessage(chatID, "Invalid option index.", nil)
			return err
		}
		optionIndex = optionIndex*10 + int(ch-'0')
	}

	// Read the temp file.
	tmpFile := filepath.Join(os.TempDir(), "ask-user-"+requestID+".json")
	data, err := os.ReadFile(tmpFile)
	if err != nil {
		_, sendErr := b.SendMessage(chatID, "Request expired or not found.", nil)
		return sendErr
	}

	var req askUserRequest
	if err := json.Unmarshal(data, &req); err != nil {
		_, sendErr := b.SendMessage(chatID, "Failed to parse request.", nil)
		return sendErr
	}

	if optionIndex < 0 || optionIndex >= len(req.Options) {
		_, sendErr := b.SendMessage(chatID, "Option index out of range.", nil)
		return sendErr
	}

	selectedText := req.Options[optionIndex]

	// Edit the message to confirm selection.
	if msgID != 0 {
		_, _, _ = b.EditMessageText("Selected: "+selectedText, &gotgbot.EditMessageTextOpts{
			ChatId:    chatID,
			MessageId: msgID,
		})
	}

	// Delete the temp file.
	_ = os.Remove(tmpFile)

	// Send the selected option to Claude.
	return enqueueGsdCommand(b, chatID, selectedText, store, mappings, cfg)
}

// enqueueGsdCommand looks up the project mapping for the channel, gets or creates
// the session, ensures the worker is started, and enqueues the message with a
// streaming callback.
//
// This is used by all GSD/option/askuser callbacks to send text to Claude.
func enqueueGsdCommand(b *gotgbot.Bot, chatID int64, text string,
	store *session.SessionStore, mappings *project.MappingStore,
	cfg *config.Config) error {

	mapping, hasMapped := mappings.Get(chatID)
	if !hasMapped {
		_, err := b.SendMessage(chatID, "No project linked. Use /project to link one first.", nil)
		return err
	}

	sess := store.GetOrCreate(chatID, mapping.Path)

	// Start worker if not already started.
	if !sess.WorkerStarted() {
		sess.SetWorkerStarted()
		callbackWg.Add(1)
		go func(s *session.Session) {
			defer callbackWg.Done()
			wCfg := session.WorkerConfig{
				AllowedPaths: []string{mapping.Path},
				SafetyPrompt: config.BuildSafetyPrompt([]string{mapping.Path}),
				FilteredEnv:  config.FilteredEnv(),
			}
			s.Worker(context.Background(), cfg.ClaudeCLIPath, wCfg)
		}(sess)
	}

	// Create streaming state and callback (no global rate limiter in callback path).
	ss := NewStreamingState(b, chatID, nil)

	qMsg := session.QueuedMessage{
		Text:   text,
		ChatID: chatID,
		Callback: func(_ int64) claude.StatusCallback {
			return CreateStatusCallback(ss)
		},
		ErrCh: make(chan error, 1),
	}

	if !sess.Enqueue(qMsg) {
		_, err := b.SendMessage(chatID, "Queue full, please wait for the current query to finish.", nil)
		return err
	}

	// Drain the error channel asynchronously.
	go func() {
		err, ok := <-qMsg.ErrCh
		if !ok || err == nil {
			// Success — check for response buttons.
			fullText := ss.AccumulatedText()
			if fullText != "" {
				maybeAttachActionKeyboard(b, chatID, fullText)
			}
			return
		}
		truncated := err.Error()
		if len(truncated) > 200 {
			truncated = truncated[:200]
		}
		_, _ = b.SendMessage(chatID, "Claude error: "+truncated, nil)
	}()

	return nil
}

// handleCallbackResume restores a saved session by session ID.
// Updates the inline keyboard message to confirm the restore.
func handleCallbackResume(b *gotgbot.Bot, _ *ext.Context, store *session.SessionStore, cfg *config.Config, chatID, msgID int64, sessionID string) error {
	if sessionID == "" {
		_, err := b.SendMessage(chatID, "Invalid session ID.", nil)
		return err
	}

	sess := store.GetOrCreate(chatID, cfg.WorkingDir)
	sess.SetSessionID(sessionID)

	// Show a short prefix of the session ID for confirmation.
	short := sessionID
	if len(short) > 8 {
		short = short[:8]
	}

	confirmText := fmt.Sprintf("Session restored: %s... — send a message to continue.", short)

	// Edit the keyboard message to remove the buttons and show the confirmation.
	if msgID != 0 {
		_, _, _ = b.EditMessageText(confirmText, &gotgbot.EditMessageTextOpts{
			ChatId:    chatID,
			MessageId: msgID,
		})
	} else {
		_, _ = b.SendMessage(chatID, confirmText, nil)
	}

	return nil
}

// handleCallbackStop stops the current running query (same as /stop).
func handleCallbackStop(b *gotgbot.Bot, store *session.SessionStore, chatID, msgID int64) error {
	sess, ok := store.Get(chatID)
	if !ok || !sess.IsRunning() {
		_, err := b.SendMessage(chatID, "No query running.", nil)
		return err
	}

	sess.Stop()

	if msgID != 0 {
		_, _, _ = b.EditMessageText("Query stopped.", &gotgbot.EditMessageTextOpts{
			ChatId:    chatID,
			MessageId: msgID,
		})
	} else {
		_, _ = b.SendMessage(chatID, "Query stopped.", nil)
	}

	return nil
}

// handleCallbackNew starts a fresh session (same as /new).
func handleCallbackNew(b *gotgbot.Bot, store *session.SessionStore, cfg *config.Config, chatID, msgID int64) error {
	sess := store.GetOrCreate(chatID, cfg.WorkingDir)

	if sess.IsRunning() {
		sess.Stop()
	}

	sess.SetSessionID("")

	if msgID != 0 {
		_, _, _ = b.EditMessageText("New session started. Previous session saved for /resume.", &gotgbot.EditMessageTextOpts{
			ChatId:    chatID,
			MessageId: msgID,
		})
	} else {
		_, _ = b.SendMessage(chatID, "New session started. Previous session saved for /resume.", nil)
	}

	return nil
}
